beacon-chain/rpc/beaconv1/state.go

```diff
@@ -56,3 +100,80 @@ func (bs *Server) GetStateFork(ctx context.Context, req *ethpb.StateRequest) (*e
 func (bs *Server) GetFinalityCheckpoints(ctx context.Context, req *ethpb.StateRequest) (*ethpb.StateFinalityCheckpointResponse, error) {
 	return nil, errors.New("unimplemented")
 }
+
+func (bs *Server) headStateRoot(ctx context.Context) ([]byte, error) {
+	stateRoot, err := bs.ChainInfoFetcher.HeadRoot(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain root: %v", err)
+	}
+	return stateRoot, nil
+}
+
+func (bs *Server) genesisStateRoot(ctx context.Context) ([]byte, error) {
+	stateRoot, err := bs.ChainStartFetcher.PreGenesisState().HashTreeRoot(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain root: %v", err)
+	}
+	return stateRoot[:], nil
+}
+
+func (bs *Server) finalizedStateRoot(ctx context.Context) ([]byte, error) {
+	var blockRoot [32]byte
+	copy(blockRoot[:], bs.ChainInfoFetcher.FinalizedCheckpt().Root)
+	state, err := bs.StateGenService.StateByRoot(ctx, blockRoot)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain state: %v", err)
+	}
+	stateRoot, err := state.HashTreeRoot(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain root: %v", err)
+	}
+	return stateRoot[:], nil
+}
+
+func (bs *Server) justifiedStateRoot(ctx context.Context) ([]byte, error) {
+	var blockRoot [32]byte
+	copy(blockRoot[:], bs.ChainInfoFetcher.CurrentJustifiedCheckpt().Root)
+	state, err := bs.StateGenService.StateByRoot(ctx, blockRoot)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain state: %v", err)
+	}
+	stateRoot, err := state.HashTreeRoot(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain root: %v", err)
+	}
+	return stateRoot[:], nil
+}
+
+func (bs *Server) stateRootByHex(ctx context.Context, stateId []byte) ([]byte, error) {
+	var stateRoot [32]byte
+	copy(stateRoot[:], stateId)
+	headState, err := bs.ChainInfoFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain head state: %v", err)
+	}
+	for _, root := range headState.StateRoots() {
+		if bytes.Equal(root, stateRoot[:]) {
+			return stateRoot[:], nil
+		}
+	}
+	return nil, status.Errorf(
+		codes.NotFound,
+		"State not found in the last %d states", len(headState.StateRoots()))
+}
+
+func (bs *Server) stateRootBySlot(ctx context.Context, slot uint64) ([]byte, error) {
+	currentSlot := bs.ChainInfoFetcher.HeadSlot()
+	if slot > currentSlot {
```

Comment:
Use TimeFetcher's `CurrentSlot()` for compare.  `HeadSlot` is not a good representation of current time due to reorgs

Reply:
Just to clarify - if there is a reorg than head slot might be less than current slot, right? In that case `slot > currentSlot` might fail even though it shouldn't. In this case it will fail on fetching the state by slot, though. I guess the latter is indeed better as we are not fetching a future slot.